src/vision/status_quo/alan_builds_a_task_scheduler.md

```diff
@@ -0,0 +1,704 @@
+            Some(guard) => _mu = guard,
+            None => return Err(...),
+        }
+
+        let request_tx;
+        match _mu.get(&id) {
+            Some(ch) => request_tx = ch,
+            None => return Err(...),
+        }
+
+        match request_tx.send(x) {
+            Ok(_) => _mu.remove(&id),
+            Err(e) => return Err(...),
+        }
+
+        Ok(())
+    }
+}
+```
+
+Alan is still not sure if `crossbeam_channel` is safe for async programming and whether he should instead use another crate [`async_std::channel`](https://docs.rs/async-std/1.9.0/async_std/channel/index.html). While `crossbeam_channel` seems to work, Alan is not confident about his choice. Disgruntled with seemingly unnecessary divergence in the community, Alan wonders why all those cool improvements had not been made back to Rust core std libraries.
+
+### **Implementing task applier**
+
+Alan implements a task applier, which simply echoes the requested message, as in Go:
+
+```go
+type EchoManager interface {
+	apply(req *EchoRequest) (string, error)
+}
+
+type echoManager struct {
+	mu sync.RWMutex
+}
+
+func (ea *echoManager) apply(req *EchoRequest) (string, error) {
+	ea.mu.Lock()
+	defer ea.mu.Unlock()
+	switch req.Kind {
+	case "create":
+		return fmt.Sprintf("SUCCESS create %q", req.Message), nil
+	case "delete":
+		return fmt.Sprintf("SUCCESS delete %q", req.Message), nil
+	default:
+		return "", fmt.Errorf("unknown request %q", req)
+	}
+}
+```
+
+Having implemented event notifier above, Alan is now somewhat familiar with Rust mutex and writes the following Rust code:
+
+```rust
+// 1st version
+use async_std::sync::RwLock;
+
+pub struct Manager {
+    mu: RwLock<()>,
+}
+
+impl Manager {
+    pub fn new() -> Self {
+        Self {
+            mu: RwLock::new(()),
+        }
+    }
+
+    pub fn apply(&self, req: &Request) -> io::Result<String> {
+        let _mu;
+        match self.mu.try_write() {
+            Some(guard) => _mu = guard,
+            None => return Err(...),
+        }
+        match req.kind.as_str() {
+            "create" => Ok(format!(
+                "SUCCESS create {}",
+                to_string(req.message.to_owned())
+            )),
+            "delete" => Ok(format!(
+                "SUCCESS delete {}",
+                to_string(req.message.to_owned())
+            )),
+            _ => Err(...),
+        }
+    }
+}
+```
+
+The code compiles and thus must be safe. However, after discussing the code with [Barbara][], Alan learns that `std::sync::Mutex` protects data from concurrent access, but `std::sync::Mutex` itselt must be also protected between threads. And this is where [`std::sync::Arc`](https://doc.rust-lang.org/std/sync/struct.Arc.html) comes in to provide safe multi-threaded access to the `Mutex`. Deeply confused, Alan made a quick fix to wrap `Mutex` with `Arc`:
+
+```rust
+// 2nd version
+use async_std::{sync::Arc, sync::RwLock};
+
+pub struct Manager {
+    mu: Arc<RwLock<()>>,
```

Comment:
What is `mu`?

Reply:
Please note this is an attempt to transliterate Go code to Rust, where it's common to define an empty `sync.Mutex` field in struct with a variable name `mu`. I will name it better for this example :)

Reply:
Ah, yes I noticed that. But I think a better approach would be to translate module by module based on functionality if you are translating from go to rust. Go mental model does not fit rust very well, if you are translating from python to rust then it may be a good idea to translate line by line but not quite a good idea for go. If you are translating, thinking in terms of how data flow in rust would ease the translation, like what data should a struct contains and such. But yeah, what you should could the the first version of translation. Maybe I am looking at the status quo differently, now it seemed like the hard part is the translation part from go to rust as the memory model is different.